app/scripts/modules/core/src/search/global/GlobalSearch.tsx

```diff
@@ -173,92 +214,120 @@ export class GlobalSearch extends React.Component<{}, IGlobalSearchState> {
-    this.setState({ showRecentItems: true });
+  private hideDropdown() {
+    this.setState({ showDropdown: false });
   }
 
   private clearFilters(result: ISearchResult) {
     ReactInjector.clusterFilterService.overrideFiltersForUrl(result);
   }
 
-  public render() {
+  private renderDropdown() {
     const {
-      showSearchResults,
-      showRecentItems,
-      showMinLengthWarning,
       query,
       querying,
+      showMinLengthWarning,
       categories
     } = this.state;
 
-    const { searchVersion } = SETTINGS;
+    if (!query) {
+      return (
+        <RecentlyViewedItems
+          limit={5}
+          Component={({ results }: IChildComponentProps) => {
+            this.resultRefs = results.map(() => []);
+
+            return (
+              <GlobalSearchRecentItems
+                categories={results}
+                onItemKeyDown={this.navigateResult}
+                onResultClick={() => {
+                  this.hideDropdown();
+                  ReactGA.event({ category: 'Global Search', action: 'Recent Item Selected' });
+                }}
+                resultRef={(categoryIndex, resultIndex, ref) => {
+                  if (this.resultRefs[categoryIndex]) {
+                    this.resultRefs[categoryIndex][resultIndex] = ref;
+                  }
+                }}
+              />
+            );
+          }}
+        />
+      );
+    }
+
+    if (querying) {
+      return (
+        <ul className="dropdown-menu" role="menu">
+          <li className="horizontal middle center spinner-section">
+            <Spinner size="small" />
+          </li>
+        </ul>
+      );
+    }
+
+    if (query.length < MIN_SEARCH_LENGTH && showMinLengthWarning) {
+      return (
+        <ul className="dropdown-menu" role="menu">
+          <li className="horizontal middle center spinner-section">
+            <span className="error-message">Please enter at least 3 characters</span>
+          </li>
+        </ul>
+      );
+    } else if (categories) {
+      return (
+        <GlobalSearchResults
+          categories={categories}
+          query={query}
+          onItemKeyDown={this.navigateResult}
+          onResultClick={(result: ISearchResult) => {
+            ReactGA.event({ category: 'Global Search', action: 'Result Selected' });
+            this.hideDropdown();
+            this.clearFilters(result);
+          }}
+          onSeeMoreClick={() => {
+            ReactGA.event({ category: 'Global Search', action: 'See more results selected' });
+          }}
+          resultRef={(categoryIndex, resultIndex, ref) => {
+            if (this.resultRefs[categoryIndex]) {
+              this.resultRefs[categoryIndex][resultIndex] = ref;
+            }
+          }}
+          seeMoreRef={(ref) => {
+            // Make sure keyboard handling works even though this isn't part of the actual results
+            this.resultRefs[categories.length] = [ref];
+          }}
+        />
+      );
+    }
```

Comment:
Totally up to you, but it might be easier to reason about and read if these different returned blocks were split into their own methods, something like:
```javascript
if (!query) {
  return this.renderRecentlyViewedItems();
}
if (querying) {
  return this.renderSpinner();
}
if (query.length < MIN_SEARCH_LENGTH && showMinLengthWarning) {
  return this.renderMinLengthWarning();
} else if (categories) {
  return this.renderCategories();
}
return null;
```

Reply:
Agreed, and I also prefer using SFC overr `renderFoo()` 
```
if (!query) {
  return <RecentlyViewedItems/>
}
if (querying) {
  return <Spinner/>
}
if (query.length < MIN_SEARCH_LENGTH && showMinLengthWarning) {
  return <MinLengthWarning/>
} else if (categories) {
  return <Categories categories={categories}/>
}
return null;
```

Reply:
I was actually going to opt for the SFC way of going at it, because that's my preference as well, but the results and recent items both touch stuff on the class instance a lot and it starts to get to a really boilerplate-tastic place where you've got *two* layers of ref/keydown/click callbacks rather than just the single layer we have right now.

Reply:
FWIW, SFC as an arrow function should have access to `this`

Reply:
you know, i don't think i've ever actually used an SFC that's scoped inside a class. i guess today is the day.